aws-amplify/docs
# Reacting to updates for esm packages (custom aloglia search)

These changes are reacting to the `unified` and `unist-util-visit` package upgrades from the #4100 PR.

These packages converted to use `esm` modules and as a result, references to these modules need to be updated to use `import` statements instead of `require`.

_Description of changes:_
- Convert the plugins using `unist-util-visit` to be async modules so that we can `await import('unist-util-visit')` instead of using `require`.
- Update `next.js` to version 12.1.4 so that `next.config.js` can use async functions and await the plugins mentioned above.
- Reacting to changes in `next` that made `patch-next-scrolling.ts` no longer work. Instead use `history.replaceState` in conjunction with next `router.replace()` to handle updating the heading hashes during scroll.

Link to amplify app: https://timngyn-custom-algolia-search.d1oosfztpocl9c.amplifyapp.com/
To test:
1. Go to https://timngyn-custom-algolia-search.d1oosfztpocl9c.amplifyapp.com/lib/datastore/getting-started/q/platform/js/
2. Scroll around the page and verify that the page is not skipping around.
3. Click through the docs site and verify that back and forward browser buttons are working with the hashes

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
